### PR TITLE
GSK-749 working on tensorflow implementation

### PR DESCRIPTION
**Pull request by rabah-khalek**
daughter of daughter of https://github.com/Giskard-AI/giskard/pull/618, https://github.com/Giskard-AI/giskard/pull/661, https://github.com/Giskard-AI/giskard/pull/662 and https://github.com/Giskard-AI/giskard/pull/664

Working on pytorch mlflow integration.

- [x] Decide on how to wrap the tensorflow models, specially those of HuggingFace
- [x] Verify implementation with the imdb [notebook](https://github.com/Giskard-AI/giskard-examples/blob/task/GSK-273_Tensorflow_model_integration_with_Adam_optimizer_IMDB_notebook/Text_classification_Using_Tensorflow_Neural_Network.ipynb)
- [x] Verify implementation with all the tensorflow unit-tests implemented in https://github.com/Giskard-AI/giskard/pull/376

### CODE REVIEW

**Comment by rabah-khalek** [comment]
The unit-test that I was planning to add, turned out to be a hugging face model, I'll defer these to another branch, and another class dedicated for transformers from hugging face as we discussed. Thereby this branch can be merged.

File: python-client/tests/tensorflow/test_mnist.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+def test_mnist():
+
+    (train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.mnist.load_data()
+
+    # Create a basic model instance
+    model = create_model()
+
+    train_labels = train_labels[:100]
+    #test_labels = test_labels[:100]
```
Comment (andreybavt) [inline python-client/tests/tensorflow/test_mnist.py]: commented code can be removed

File: python-client/tests/tensorflow/test_mnist.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    train_images = train_images[:100].reshape(-1, 28 * 28) / 255.0
+    #test_images = test_images[:100].reshape(-1, 28 * 28) / 255.0
```
Comment (andreybavt) [inline python-client/tests/tensorflow/test_mnist.py]: commented code can be removed

File: python-client/tests/tensorflow/test_mnist.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    with requests_mock.Mocker() as m:
+        m.register_uri(requests_mock.POST, artifact_url_pattern)
+        m.register_uri(requests_mock.POST, models_url_pattern)
+        m.register_uri(requests_mock.GET, settings_url_pattern)
+
+        url = "http://giskard-host:12345"
+        token = "SECRET_TOKEN"
+        client = GiskardClient(url, token)
+        # enron = client.create_project('test-project', "Email Classification", "Email Classification")
```
Comment (andreybavt) [inline python-client/tests/tensorflow/test_mnist.py:71]: commented code can be removed

File: python-client/tests/tensorflow/test_text_classification_tfhub.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import tensorflow_text
+
+from giskard.client.giskard_client import GiskardClient
+from giskard import TensorFlowModel, Dataset
+
+import requests_mock
+import re
+import tests.utils
+
+#tfhub_handle_encoder = 'https://tfhub.dev/tensorflow/small_bert/bert_en_uncased_L-4_H-512_A-8/1'
```
Comment (andreybavt) [inline python-client/tests/tensorflow/test_text_classification_tfhub.py]: commented code can be removed

File: python-client/tests/tensorflow/test_text_classification_tfhub.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    with requests_mock.Mocker() as m:
+        m.register_uri(requests_mock.POST, artifact_url_pattern)
+        m.register_uri(requests_mock.POST, models_url_pattern)
+        m.register_uri(requests_mock.GET, settings_url_pattern)
+
+        url = "http://giskard-host:12345"
+        token = "SECRET_TOKEN"
+        client = GiskardClient(url, token)
+        # enron = client.create_project('test-project', "Email Classification", "Email Classification")
```
Comment (andreybavt) [inline python-client/tests/tensorflow/test_text_classification_tfhub.py:101]: commented code can be removed

**Review by andreybavt — CHANGES_REQUESTED** [review]

**Comment by rabah-khalek** [comment]
Done.

File: python-client/giskard/core/model.py [hunk 6]
```diff
@@ -289,15 +289,38 @@ def __init__(self,
         self.data_preprocessing_function = data_preprocessing_function
         self.model_postprocessing_function = model_postprocessing_function
 
+    def post_processing(self, raw_prediction):
```
Comment (andreybavt) [inline python-client/giskard/core/model.py:292]: to make it less confusing for an end user which post processing function he has to override, I'd make this method private by adding `_` prefix

File: python-client/giskard/core/model.py [hunk 7]
```diff
@@ -289,15 +289,38 @@ def __init__(self,
...
+        warning_message = f"\nYour binary classification model prediction is of the shape {raw_prediction.shape}. \n" + \
+                          f"In Giskard we expect the shape {(raw_prediction.shape[0], 2)} for binary classification models. \n" + \
+                          "We automatically infered the second class prediction but please make sure that \n" + \
```
Comment (andreybavt) [inline python-client/giskard/core/model.py:295]: infe`rr`ed

File: python-client/giskard/core/model.py [hunk 8]
```diff
@@ -289,15 +289,38 @@ def __init__(self,
         self.data_preprocessing_function = data_preprocessing_function
         self.model_postprocessing_function = model_postprocessing_function
 
+    def post_processing(self, raw_prediction):
+        warning_message = f"\nYour binary classification model prediction is of the shape {raw_prediction.shape}. \n" + \
```
Comment (andreybavt) [inline python-client/giskard/core/model.py:293]: nitpick: `warning_message` can be initialized inside `if is_binary_classification`

**Review by andreybavt — CHANGES_REQUESTED** [review]
mostly last minor comments

**Pull request merged**